libraries/Microsoft.Bot.Builder.Ai/TranslationMiddleware.cs

```diff
@@ -0,0 +1,177 @@
+        {
+            if (nativeLanguages == null || nativeLanguages.Count() == 0)
+                throw new ArgumentNullException(nameof(nativeLanguages));
+            this._nativeLanguages = nativeLanguages;
+            if (string.IsNullOrEmpty(translatorKey))
+                throw new ArgumentNullException(nameof(translatorKey));
+            this._translator = new Translator(translatorKey);
+            this._patterns = patterns ?? throw new ArgumentNullException(nameof(patterns));
+        }
+
+        /// <summary>
+        /// Constructor for developer defined detection of user messages and using templates
+        /// </summary>
+        /// <param name="nativeLanguages">List of languages supported by your app</param>
+        /// <param name="translatorKey"></param>
+        /// <param name="patterns">Dictionary with language as a key and list of patterns as value</param>
+        /// <param name="getUserLanguage">Delegate for getting the user language</param>
+        /// <param name="setUserLanguage">Delegate for setting the user language, returns true if the language was changed (implements logic to change language by intercepting the message)</param>
+        public TranslationMiddleware(string[] nativeLanguages, string translatorKey, Dictionary<string, List<string>> patterns, Func<ITurnContext, string> getUserLanguage, Func<ITurnContext, Task<bool>> setUserLanguage)
+        {
+            if (nativeLanguages == null || nativeLanguages.Count() == 0)
+                throw new ArgumentNullException(nameof(nativeLanguages));
+            this._nativeLanguages = nativeLanguages;
+            if (string.IsNullOrEmpty(translatorKey))
+                throw new ArgumentNullException(nameof(translatorKey));
+            this._translator = new Translator(translatorKey);
+            this._patterns = patterns ?? throw new ArgumentNullException(nameof(patterns));
+            this._patterns = patterns;
+            this._getUserLanguage = getUserLanguage ?? throw new ArgumentNullException(nameof(getUserLanguage));
+            this._setUserLanguage = setUserLanguage ?? throw new ArgumentNullException(nameof(setUserLanguage)); 
+        }
+
+        /// <summary>
+        /// Incoming activity
+        /// </summary>
+        /// <param name="context"></param>
+        /// <param name="next"></param>
+        /// <returns></returns>       
+        public async Task OnProcessRequest(ITurnContext context, MiddlewareSet.NextDelegate next)
+        {
+            IMessageActivity message = context.Activity.AsMessageActivity();
+            if (message != null)
+            {
+                if (!String.IsNullOrWhiteSpace(message.Text))
+                {
+
+                    var languageChanged = false;
+
+                    if (_setUserLanguage != null)
+                    {
+                        languageChanged = await _setUserLanguage(context);
```

Comment:
I wouldn't expect a method called "SetUserLanguage" to return true/false. The names makes that behavior a surprise, give how "Set" methods normally work in C#.

Reply:
changed its name to isUserLanguageChanged